python/pyspark/sql/worker/plan_data_source_read.py

```diff
@@ -54,17 +54,31 @@
 
 
 def records_to_arrow_batches(
-    output_iter: Iterator[Tuple],
+    output_iter: Union[Iterator[Tuple], Iterator[pa.RecordBatch]],
     max_arrow_batch_size: int,
     return_type: StructType,
     data_source: DataSource,
 ) -> Iterable[pa.RecordBatch]:
     """
-    Convert an iterator of Python tuples to an iterator of pyarrow record batches.
+    First check if the iterator yields pyarrow record batched, if so, yield them directly.
+    Otherwise, convert an iterator of Python tuples to an iterator of pyarrow record batches.
+    For each Python tuple, check the types of each field and append it to the records batch.
+    """
 
-    For each python tuple, check the types of each field and append it to the records batch.
+    # Create a copy of the iterator using itertools.tee
+    output_iter, output_iter_copy = tee(output_iter)
```

Comment:
It might be inefficient to create a copy of the entire `output_iter` here.

Reply:
Good point. Any tips?